### PR TITLE
Some changes to hypot

### PR DESCRIPTION
**Pull request by andreasnoack**
Following up on https://github.com/JuliaLang/julia/pull/15552.
1. Use the julia implementation of `hypot`. On my MacBook it is more that 2x faster than the libm version.
2. Loosen signature to allow for complex input
3. Make a few changes to make sure that the return type is correctly inferred for more exotic types. (Hopefully, one day the dead branches that cause type instability can be eliminated such that it's not necessary to convert the return elements in the dead branches.)
4. Add an arbitrary argument version of `hypot`. This supersedes #15552. ~~(just realized that doc update is missing. I'll push that tomorrow.)~~

### CODE REVIEW

File: test/math.jl [hunk 1]
```diff
@@ -177,6 +179,11 @@ for T in (Float32, Float64)
     @test isinf(log1p(-one(T)))
     @test isnan(log1p(convert(T,NaN)))
     @test_throws DomainError log1p(convert(T,-2.0))
+    @test hypot(T(0), T(0)) == T(0)
```
Comment (tkelman) [inline test/math.jl]: `===` to test for type equality too?

Comment (andreasnoack) [inline test/math.jl]: Yes. Just did that.

File: base/math.jl [hunk 2]
```diff
@@ -146,36 +146,51 @@ sqrt(x::Float32) = box(Float32,sqrt_llvm(unbox(Float32,x)))
...
+hypot(x::Number, y::Number) = hypot(promote(x, y)...)
+function hypot{T<:Number}(x::T, y::T)
+    ax = abs(x)
+    ay = abs(y)
+    if ax < ay
+        ax, ay = ay, ax
     end
-    if x == 0
-        r = y/one(x)
+    if ax == 0
```
Comment (musmo) [inline base/math.jl]: is comparing to `0` sufficient or do we need `if ax == zero(ax)` ?

Comment (nalimilan) [inline base/math.jl]: For comparisons, you don't need to call `zero`. It's only useful when you want a specific type.

File: base/math.jl [hunk 3]
```diff
@@ -146,36 +146,51 @@ sqrt(x::Float32) = box(Float32,sqrt_llvm(unbox(Float32,x)))
...
-        end
+        r = ay / ax
+    end
+
+    rr = ax * sqrt(1 + r * r)
+
+    # Use type of rr to make sure that return type is the same for
+    # all branches
+    if isnan(r)
+        isinf(ax) && return typeof(rr)(Inf)
```
Comment (nalimilan) [inline base/math.jl]: `typeof(rr)(Inf)` would better be written `oftype(rr, Inf)`. Same for `typeof(rr)(r)`.

Comment (andreasnoack) [inline base/math.jl]: Thanks. Updated.

**Pull request merged**